### ISSUE TITLE
Binary segmentation does not work well because the loss is always zero. (Reproduction code available)

### ISSUE THREAD

**Issue by apiss2** [open-mmlab/mmsegmentation#2201]
I tried to do a binary segmentation on a dataset that I defined myself, but the CrossEntropyLoss is always around 0 and the training does not progress.
I get a model that always predicts positive for all of the images.

Is there something wrong with the config settings?

Attached below is the code to reproduce this in google colab.

1. install mmsegmentation and download dataset
``` bash
!pip install -U openmim
!mim install mmcv-full
!git clone https://github.com/open-mmlab/mmsegmentation.git
!pip install -v -e ./mmsegmentation
!if [ ! -d train ]; then curl -L -O https://github.com/mitmul/chainer-handson/releases/download/SegmentationDataset/train.zip && unzip train.zip && rm -rf train.zip; fi
!if [ ! -d val ]; then curl -L -O https://github.com/mitmul/chainer-handson/releases/download/SegmentationDataset/val.zip && unzip val.zip && rm -rf val.zip; fi
```

2. visualize image in this dataset
``` python
import matplotlib.pyplot as plt
import numpy as np

from PIL import Image

img = np.asarray(Image.open('train/image/000.png'))
label = np.asarray(Image.open('train/label/000.png'))
print('label unique value:', np.unique(label))

fig, axes = plt.subplots(1, 2)
axes[0].set_axis_off()
axes[0].imshow(img, cmap='gray')
axes[1].set_axis_off()
axes[1].imshow(label, cmap='gray')
plt.show()
```
![image](https://japan-medical-ai.github.io/medical-ai-course-materials/_images/notebooks_05_Image_Segmentation_9_0.png)

3. Setup a Config other than dataset
``` python
from mmcv import Config

config = Config.fromfile('/content/mmsegmentation/configs/_base_/models/lraspp_m-v3-d8.py')
config.merge_from_dict(Config.fromfile('/content/mmsegmentation/configs/_base_/default_runtime.py'))
config.merge_from_dict(Config.fromfile('/content/mmsegmentation/configs/_base_/schedules/schedule_20k.py'))
config.merge_from_dict(dict(samples_per_gpu=4, workers_per_gpu=2))
config.merge_from_dict(dict(type='IterBasedRunner', max_iters=20000))

config.model.decode_head.num_classes = 1
config.model.decode_head.out_channels = 1
config.model.decode_head.threshold = 0.3
config.model.decode_head.loss_decode.avg_non_ignore = True
config.model.backbone.norm_cfg.type = "BN"
config.model.decode_head.norm_cfg.type = "BN"
config.model.decode_head.ignore_index = 0
config.optimizer.lr = 0.0001
config.checkpoint_config.interval = 200
config.evaluation.interval = 200
config.runner.max_iters = 2000
```

4. Update Config to match the downloaded dataset
``` python
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53],
    std=[58.395, 57.12, 57.375],
    to_rgb=True
)
img_scale = (520, 520)
crop_size = (480, 480)

train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations'),
    dict(type='Resize', img_scale=img_scale, ratio_range=(0.8, 1.2)),
    dict(type='RandomCrop', crop_size=crop_size, cat_max_ratio=0.75),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='PhotoMetricDistortion'),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size=crop_size, pad_val=0, seg_pad_val=0),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_semantic_seg']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=img_scale,
        # img_ratios=[0.5, 0.75, 1.0, 1.25, 1.5, 1.75],
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ]
        )
]
dataset_type = 'CustomDataset'
data_root = './'
palette = [[0, 0, 0], [255, 0, 0]]
classes = ('background', 'target')
data = dict(
    samples_per_gpu=4,
    workers_per_gpu=2,
    train=dict(
        type=dataset_type,
        data_root=data_root,
        img_dir='train/image',
        ann_dir='train/label',
        img_suffix='.png',
        seg_map_suffix='.png',
        classes=classes,
        palette=palette,
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        data_root=data_root,
        img_dir='val/image',
        ann_dir='val/label',
        img_suffix='.png',
        seg_map_suffix='.png',
        classes=classes,
        palette=palette,
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        data_root=data_root,
        img_dir='val/image',
        ann_dir='val/label',
        img_suffix='.png',
        seg_map_suffix='.png',
        classes=classes,
        palette=palette,
        pipeline=test_pipeline
        )
)
```

5. Dump config
```
config.merge_from_dict({"data": data})
config.dump('./config.py')
```
6. training
```
!python mmsegmentation/tools/train.py ./config.py
```

Once learning begins, the following output is obtained
where decode.loss_ce and decode.acc_seg are always near 0.
```
2022-10-18 00:56:14,680 - mmseg - INFO - workflow: [('train', 1)], max: 2000 iters
2022-10-18 00:56:14,680 - mmseg - INFO - Checkpoints will be saved to /content/work_dirs/config by HardDiskBackend.
2022-10-18 00:56:30,109 - mmseg - INFO - Iter [50/2000]	lr: 1.000e-04, eta: 0:09:58, time: 0.307, data_time: 0.017, memory: 7433, decode.loss_ce: -0.0020, decode.acc_seg: 0.0000, loss: -0.0020
2022-10-18 00:56:43,163 - mmseg - INFO - Iter [100/2000]	lr: 1.000e-04, eta: 0:08:59, time: 0.261, data_time: 0.056, memory: 7433, decode.loss_ce: -0.0002, decode.acc_seg: 0.0000, loss: -0.0002
2022-10-18 00:56:56,596 - mmseg - INFO - Iter [150/2000]	lr: 1.000e-04, eta: 0:08:36, time: 0.269, data_time: 0.062, memory: 7433, decode.loss_ce: -0.0006, decode.acc_seg: 0.0000, loss: -0.0006
2022-10-18 00:57:09,771 - mmseg - INFO - Saving checkpoint at 200 iterations
2022-10-18 00:57:09,953 - mmseg - INFO - Iter [200/2000]	lr: 1.000e-04, eta: 0:08:17, time: 0.268, data_time: 0.056, memory: 7433, decode.loss_ce: 0.0001, decode.acc_seg: 0.0000, loss: 0.0001
[                                                  ] 0/26, elapsed: 0s, ETA:/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:1960: UserWarning: nn.functional.sigmoid is deprecated. Use torch.sigmoid instead.
  warnings.warn("nn.functional.sigmoid is deprecated. Use torch.sigmoid instead.")
[>>] 26/26, 12.2 task/s, elapsed: 2s, ETA:     0s2022-10-18 00:57:12,081 - mmseg - INFO - per class results:
2022-10-18 00:57:12,082 - mmseg - INFO - 
+------------+------+-------+
|   Class    | IoU  |  Acc  |
+------------+------+-------+
| background | 0.0  |  0.0  |
|   target   | 1.77 | 100.0 |
+------------+------+-------+
2022-10-18 00:57:12,082 - mmseg - INFO - Summary:
2022-10-18 00:57:12,083 - mmseg - INFO - 
+------+------+------+
| aAcc | mIoU | mAcc |
+------+------+------+
| 1.77 | 0.88 | 50.0 |
+------+------+------+
```
7. test
Outputting an image as follows will result in an image with the entire image painted red.
```
!python mmsegmentation/tools/test.py ./config.py work_dirs/config/latest.pth --work-dir work_dirs/config --show-dir ./result
```

**Comment by MengzhangLI** [comment]
Could you try to set `num_classes= 2` to see whether it is caused by wrong config setting. You can follow similar dataset setting [here](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/datasets/drive.py#L17) and related [config](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/_base_/datasets/drive.py).

**Comment by apiss2** [comment]
Thanks for your quick response!
I followed your advice and changed the config to `num_classes=2`, but it did not improve.

However, I found that removing the changes to out_channels and threshold, in addition to the above changes, resulted in successful learning.
Thank you!


Is it wrong to set out_channels to 1 for binary segmentation?
It seems redundant to set out_channels to 2.

**Comment by MeowZheng** [comment]
please set 
```python
config.model.decode_head.loss_decode.use_sigmoid=True
```

**Comment by apiss2** [comment]
By setting use_sigmoid=True, we were able to successfully learn even when out_channels=1.
Appreciate your support!

**Comment by chefkrym** [comment]
> Thanks for your quick response! I followed your advice and changed the config to `num_classes=2`, but it did not improve.
> 
> However, I found that removing the changes to out_channels and threshold, in addition to the above changes, resulted in successful learning. Thank you!
> 
> Is it wrong to set out_channels to 1 for binary segmentation? It seems redundant to set out_channels to 2.

Hello @apiss2 Could you kindly share your wiorking code?
